File: 백준/Gold/14891. 톱니바퀴/톱니바퀴.py
gear = [list(map(int, input())) for _ in range(4)]
rot = [0, 0, 0, 0] # rotation index


for _ in range(int(input())):
    i, d = map(int, input().split()) # number, direction
    i, d = i-1, -d # index, delta
    turn = [(i, d)]

    # right part
    nd = d
    for ni in range(i+1, 4):
        nd = -nd
        if gear[ni-1][rot[ni-1]-6] != gear[ni][rot[ni]-2]:
            turn.append((ni, nd))
        else:
            break

    # left part
    nd = d
    for ni in range(i-1, -1, -1):
        nd = -nd
        if gear[ni+1][rot[ni+1]-2] != gear[ni][rot[ni]-6]:
            turn.append((ni, nd))
        else:
            break
        
    # turn
    for i, d in turn:
        rot[i] = (rot[i] + 8 + d) % 8


ans = 0
for i in range(4):
    ans += gear[i][rot[i]] * 2**i
print(ans)

File: 백준/Gold/14891. 톱니바퀴/톱니바퀴.js

function turn(num, dir) {
  const delta = new Array(4).fill(0);
  delta[num] -= dir;
  
  // right part
  let nDir = dir;
  for (let i = num; i < 3; i++) {
    if (gear[i][(idx[i] + 2) % 8] !== gear[i+1][(idx[i+1] + 6) % 8]) {
      delta[i+1] += nDir;
      nDir = -nDir;
    } else {
      break;
    }
  }
  
  // left part
  nDir = dir;
  for (let i = num; i > 0; i--) {
    if (gear[i][(idx[i] + 6) % 8] !== gear[i-1][(idx[i-1] + 2) % 8]) {
      delta[i-1] += nDir;
      nDir = -nDir;
    } else {
      break;
    }
  }
  
  // turn
  for (let i = 0; i < 4; i++) {
    idx[i] = (idx[i] + delta[i] + 8) % 8;
  }
}


const input = require('fs').readFileSync(process.platform==='linux'?0:'input.txt').toString().trim().split('\n');
const gear = input.slice(0, 4).map(x=>x.split('').map(Number));
const idx = Array(4).fill(0);

for (let i=5; i<input.length; i++) {
  const [num, dir] = input[i].split(' ').map(Number);
  turn(num - 1, dir);
}

let ans = 0;
for (let i=0; i<4; i++) {
  ans += gear[i][idx[i]] * 2 ** i;
}
console.log(ans);